File: Sources/SkillManagerCore/UsageAnalyzer.swift
import Foundation

public final class UsageAnalyzer: @unchecked Sendable {
    private let fileManager: FileManager
    private let homeURL: URL
    private let maxLogBytes: UInt64
    private let maxLogFiles: Int

    private struct UsageSearchText {
        var text: String
        var kind: UsageEvidenceKind
        var agent: SkillAgent
        var detail: String
    }

    public init(
        fileManager: FileManager = .default,
        homeURL: URL = FileManager.default.homeDirectoryForCurrentUser,
        maxLogBytes: UInt64 = 512 * 1_024,
        maxLogFiles: Int = 300
    ) {
        self.fileManager = fileManager
        self.homeURL = homeURL
        self.maxLogBytes = maxLogBytes
        self.maxLogFiles = maxLogFiles
    }

    public convenience init(
        fileManager: FileManager = .default,
        homeURL: URL = FileManager.default.homeDirectoryForCurrentUser,
        maxLogBytes: UInt64 = 512 * 1_024
    ) {
        self.init(
            fileManager: fileManager,
            homeURL: homeURL,
            maxLogBytes: maxLogBytes,
            maxLogFiles: 300
        )
    }

    public func analyzeSkillUsage(skillNames: [String]) -> [String: UsageHit] {
        let terms = normalizedTerms(from: skillNames)
        return analyzeUsage(terms: terms)
    }

    public func analyzeSkillUsage(skills: [SkillRecord]) -> [String: UsageHit] {
        let terms = normalizedPathTerms(from: skills)
        return analyzeUsage(terms: terms)
    }

    private func analyzeUsage(terms: [String: [String]]) -> [String: UsageHit] {
        guard !terms.isEmpty else { return [:] }

        let byteTerms = terms.mapValues { variants in
            variants.compactMap { $0.data(using: .utf8) }
        }

        var hits: [String: UsageHit] = [:]
        for logURL in sessionLogURLs() {
            let modifiedAt = (try? logURL.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate) ?? Date.distantPast
            for (original, evidence) in matchedSkillEvidence(in: logURL, terms: terms, byteTerms: byteTerms, modifiedAt: modifiedAt) {
                var hit = hits[original] ?? UsageHit()
                hit.count += evidence.count
                if hit.lastUsedAt == nil || modifiedAt > (hit.lastUsedAt ?? Date.distantPast) {
                    hit.lastUsedAt = modifiedAt
                }
                hit.evidence.append(contentsOf: evidence)
                hit.evidence.sort {
                    ($0.occurredAt ?? .distantPast) > ($1.occurredAt ?? .distantPast)
                }
                if hit.evidence.count > 20 {
                    hit.evidence = Array(hit.evidence.prefix(20))
                }
                hits[original] = hit
            }
        }
        return hits
    }

    private func matchedSkillEvidence(
        in logURL: URL,
        terms: [String: [String]],
        byteTerms: [String: [Data]],
        modifiedAt: Date
    ) -> [String: [UsageEvidence]] {
        guard let data = try? Data(contentsOf: logURL, options: [.mappedIfSafe]), !data.isEmpty else {
            return [:]
        }

        var result: [String: [UsageEvidence]] = [:]
        for (lineIndex, rawLine) in data.split(separator: 10, omittingEmptySubsequences: true).enumerated() {
            let line = Data(rawLine)
            guard line.containsASCII("\"name\":\"Skill\"") ||
                    line.containsASCII("\"name\": \"Skill\"") ||
                    line.containsASCII("\"name\":\"loadSkill\"") ||
                    line.containsASCII("\"name\": \"loadSkill\"") ||
                    line.containsASCII("\"name\":\"load_skill\"") ||
                    line.containsASCII("\"name\": \"load_skill\"") ||
                    line.containsASCII("SKILL.md") ||
                    line.containsASCII("/skills/") else {
                continue
            }
            guard let object = try? JSONSerialization.jsonObject(with: line) else {
                continue
            }

            var lineMatches: [String: UsageEvidence] = [:]

            if let skillName = findSkillToolUse(in: object),
               let original = originalName(for: skillName, terms: terms) {
                let isClaude = logURL.path.contains("/.claude/projects/")
                let kind: UsageEvidenceKind = isClaude ? .claudeSkillTool : .codexDirectLoad
                lineMatches[original] = usageEvidence(
                    original: original,
                    logURL: logURL,
                    lineIndex: lineIndex,
                    kind: kind,
                    agent: isClaude ? .claude : .codex,
                    modifiedAt: modifiedAt,
                    detail: kind.label
                )
            }

            for event in codexToolSearchTexts(from: object) {
                if !event.text.contains("/") && !event.text.contains("SKILL.md"),
                   let original = originalName(for: event.text, terms: terms) {
                    lineMatches[original] = usageEvidence(
                        original: original,
                        logURL: logURL,
                        lineIndex: lineIndex,
                        kind: event.kind,
                        agent: event.agent,
                        modifiedAt: modifiedAt,
                        detail: event.detail
                    )
                    continue
                }
                guard event.text.contains("SKILL.md") || event.text.contains("/skills/") else {
                    continue
                }
                let searchData = Data(event.text.utf8)
                for (original, variants) in byteTerms {
                    if variants.contains(where: { searchData.range(of: $0) != nil }) {
                        lineMatches[original] = usageEvidence(
                            original: original,
                            logURL: logURL,
                            lineIndex: lineIndex,
                            kind: event.kind,
                            agent: event.agent,
                            modifiedAt: modifiedAt,
                            detail: event.detail
                        )
                    }
                }
            }

            guard !lineMatches.isEmpty else {
                continue
            }
            for (original, evidence) in lineMatches {
                result[original, default: []].append(evidence)
            }
        }

        return result
    }

    private func usageEvidence(
        original: String,
        logURL: URL,
        lineIndex: Int,
        kind: UsageEvidenceKind,
        agent: SkillAgent,
        modifiedAt: Date,
        detail: String
    ) -> UsageEvidence {
        let idSource = "\(original)|\(logURL.path)|\(lineIndex)|\(kind.rawValue)"
        let id = idSource.data(using: .utf8)?.base64EncodedString() ?? idSource
        return UsageEvidence(
            id: id,
            skillName: original,
            agent: agent,
            kind: kind,
            sessionPath: logURL.standardizedFileURL.path,
            occurredAt: modifiedAt,
            detail: detail
        )
    }

    private func findSkillToolUse(in value: Any) -> String? {
        if let dictionary = value as? [String: Any] {
            if dictionary["type"] as? String == "tool_use",
               dictionary["name"] as? String == "Skill",
               let input = dictionary["input"] as? [String: Any],
               let skill = input["skill"] as? String {
                return skill
            }

            for child in dictionary.values {
                if let found = findSkillToolUse(in: child) {
                    return found
                }
            }
        }

        if let array = value as? [Any] {
            for child in array {
                if let found = findSkillToolUse(in: child) {
                    return found
                }
            }
        }

        return nil
    }

    private func codexToolSearchTexts(from value: Any) -> [UsageSearchText] {
        guard let dictionary = value as? [String: Any],
              dictionary["type"] as? String == "response_item",
              let payload = dictionary["payload"] as? [String: Any],
              let payloadType = payload["type"] as? String else {
            return []
        }

        switch payloadType {
        case "function_call":
            let name = payload["name"] as? String ?? ""
            if isDirectSkillToolName(name), let skillName = skillNameFromToolPayload(payload) {
                return [UsageSearchText(text: skillName, kind: .codexDirectLoad, agent: .codex, detail: "Codex \(name)")]
            }
            guard isSearchableToolCallName(name) else { return [] }
            return stringValues(in: payload, keys: ["arguments", "input"]).map {
                UsageSearchText(text: $0, kind: .codexSkillRead, agent: .codex, detail: "Codex \(name) read")
            }

        case "custom_tool_call":
            let name = payload["name"] as? String ?? ""
            if isDirectSkillToolName(name), let skillName = skillNameFromToolPayload(payload) {
                return [UsageSearchText(text: skillName, kind: .codexDirectLoad, agent: .codex, detail: "Codex \(name)")]
            }
            guard isSearchableToolCallName(name) else { return [] }
            return stringValues(in: payload, keys: ["arguments", "input"]).map {
                UsageSearchText(text: $0, kind: .codexSkillRead, agent: .codex, detail: "Codex \(name) read")
            }

        default:
            return []
        }
    }

    private func isSearchableToolCallName(_ name: String) -> Bool {
        [
            "exec_command",
            "read_mcp_resource",
            "open"
        ].contains(name)
    }

    private func isDirectSkillToolName(_ name: String) -> Bool {
        ["Skill", "loadSkill", "load_skill"].contains(name)
    }

    private func skillNameFromToolPayload(_ payload: [String: Any]) -> String? {
        for value in stringValues(in: payload, keys: ["arguments", "input"]) {
            if let data = value.data(using: .utf8),
               let object = try? JSONSerialization.jsonObject(with: data),
               let skill = skillNameFromJSONObject(object) {
                return skill
            }
            let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
            if !trimmed.isEmpty, !trimmed.contains("/") {
                return trimmed
            }
        }
        return nil
    }

    private func skillNameFromJSONObject(_ object: Any) -> String? {
        if let dictionary = object as? [String: Any] {
            for key in ["skill", "skillName", "name"] {
                if let value = dictionary[key] as? String {
                    return value
                }
            }
        }
        return nil
    }

    private func stringValues(in dictionary: [String: Any], keys: [String]) -> [String] {
        keys.compactMap { key in
            if let value = dictionary[key] as? String {
                return value
            }
            if let value = dictionary[key],
               JSONSerialization.isValidJSONObject(value),
               let data = try? JSONSerialization.data(withJSONObject: value),
               let text = String(data: data, encoding: .utf8) {
                return text
            }
            return nil
        }
    }

    private func originalName(for observedSkillName: String, terms: [String: [String]]) -> String? {
        let normalized = observedSkillName
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased()
        guard !normalized.isEmpty else { return nil }

        for (original, variants) in terms {
            if original.lowercased() == normalized {
                return original
            }
            if variants.contains(where: { $0.lowercased() == normalized }) {
                return original
            }
        }

        return nil
    }

    public func sessionLogURLs() -> [URL] {
        var urls: [URL] = []
        urls.append(contentsOf: codexRecentSessionLogs(days: 120))
        urls.append(contentsOf: shallowLogs(in: homeURL.appendingPathComponent(".codex/archived_sessions")))
        urls.append(contentsOf: claudeProjectLogs())

        return Array(urls.sorted { lhs, rhs in
            let lDate = (try? lhs.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate) ?? .distantPast
            let rDate = (try? rhs.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate) ?? .distantPast
            return lDate > rDate
        }.prefix(maxLogFiles))
    }

    public func sessionRootAudits() -> [UsageSessionRootAudit] {
        let logs = sessionLogURLs()
        let roots: [(URL, SkillAgent)] = [
            (homeURL.appendingPathComponent(".codex/sessions"), .codex),
            (homeURL.appendingPathComponent(".codex/archived_sessions"), .codex),
            (homeURL.appendingPathComponent(".claude/projects"), .claude)
        ]

        let standardizedLogs = logs.map { $0.standardizedFileURL.path }
        return roots.map { root, agent in
            let standardizedRoot = root.standardizedFileURL.path
            return UsageSessionRootAudit(
                path: standardizedRoot,
                agent: agent,
                exists: fileManager.fileExists(atPath: standardizedRoot),
                logCount: standardizedLogs.filter { $0.hasPrefix(standardizedRoot) }.count
            )
        }
    }

    private func codexRecentSessionLogs(days: Int) -> [URL] {
        let root = homeURL.appendingPathComponent(".codex/sessions")
        let calendar = Calendar(identifier: .gregorian)
        let today = Date()
        var urls: [URL] = []

        for offset in 0..<days {
            guard let date = calendar.date(byAdding: .day, value: -offset, to: today) else { continue }
            let components = calendar.dateComponents([.year, .month, .day], from: date)
            guard let year = components.year,
                  let month = components.month,
                  let day = components.day else {
                continue
            }
            let directory = root
                .appendingPathComponent(String(format: "%04d", year))
                .appendingPathComponent(String(format: "%02d", month))
                .appendingPathComponent(String(format: "%02d", day))
            urls.append(contentsOf: shallowLogs(in: directory))
        }

        return urls
    }

    private func claudeProjectLogs() -> [URL] {
        let root = homeURL.appendingPathComponent(".claude/projects")
        guard let projects = try? fileManager.contentsOfDirectory(
            at: root,
            includingPropertiesForKeys: [.isDirectoryKey],
            options: [.skipsHiddenFiles, .skipsPackageDescendants]
        ) else {
            return []
        }

        return projects
            .prefix(120)
            .filter { isDirectory($0) }
            .flatMap { shallowLogs(in: $0).prefix(20) }
    }

    private func shallowLogs(in directory: URL) -> [URL] {
        guard fileManager.fileExists(atPath: directory.path),
              let files = try? fileManager.contentsOfDirectory(
                at: directory,
                includingPropertiesForKeys: [.isRegularFileKey, .fileSizeKey, .contentModificationDateKey],
                options: [.skipsHiddenFiles, .skipsPackageDescendants]
              ) else {
            return []
        }

        return files.filter { url in
            guard isSearchableLog(url),
                  let values = try? url.resourceValues(forKeys: [.isRegularFileKey, .fileSizeKey]),
                  values.isRegularFile == true else {
                return false
            }
            return UInt64(values.fileSize ?? 0) <= maxLogBytes
        }
    }

    private func isDirectory(_ url: URL) -> Bool {
        (try? url.resourceValues(forKeys: [.isDirectoryKey]).isDirectory) == true
    }

    private func normalizedTerms(from skillNames: [String]) -> [String: [String]] {
        var result: [String: Set<String>] = [:]
        for name in skillNames {
            let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
            guard trimmed.count >= 3 else { continue }
            var variants: Set<String> = [trimmed.lowercased()]
            variants.insert(trimmed.replacingOccurrences(of: " ", with: "-").lowercased())
            variants.insert(trimmed.replacingOccurrences(of: "_", with: "-").lowercased())
            result[trimmed] = variants
        }
        return result.mapValues { Array($0) }
    }

    private func normalizedPathTerms(from skills: [SkillRecord]) -> [String: [String]] {
        var result: [String: Set<String>] = [:]

        for skill in skills {
            let folder = URL(fileURLWithPath: skill.path).lastPathComponent
            let skillMarkdown = URL(fileURLWithPath: skill.path).appendingPathComponent("SKILL.md").path
            var variants: Set<String> = [
                skill.path,
                skillMarkdown,
                "/.agents/skills/\(folder)/",
                "/.agents/skills/\(folder)/SKILL.md",
                "/.codex/skills/\(folder)/",
                "/.codex/skills/\(folder)/SKILL.md",
                "/.claude/skills/\(folder)/",
                "/.claude/skills/\(folder)/SKILL.md"
            ]

            if !skill.relativePath.isEmpty {
                variants.insert(skill.relativePath)
                variants.insert("/\(skill.relativePath)/SKILL.md")
            }

            result[skill.name] = variants
        }

        return result.mapValues { Array($0) }
    }

    private func isSearchableLog(_ url: URL) -> Bool {
        let ext = url.pathExtension.lowercased()
        return ["jsonl", "json", "log", "txt", "md"].contains(ext)
    }
}

private extension Data {
    func containsASCII(_ text: String) -> Bool {
        guard let needle = text.data(using: .utf8), !needle.isEmpty else {
            return false
        }
        return range(of: needle) != nil
    }
}

File: packages/core/src/usage-analyzer.ts
import fs from "node:fs";
import path from "node:path";

import { isSearchableLog, isoFromDate, pathExists, stableId } from "./path-utils.js";
import type { SkillAgent, SkillRecord, UsageEvidence, UsageEvidenceAudit, UsageEvidenceKind, UsageHit, UsageSessionRootAudit } from "./types.js";

interface UsageSearchText {
  text: string;
  kind: UsageEvidenceKind;
  agent: SkillAgent;
  detail: string;
}

interface ShallowLogScan {
  paths: string[];
  oversizedCount: number;
}

interface SessionRootScan extends ShallowLogScan {
  path: string;
  agent: SkillAgent;
  exists: boolean;
  timeWindowDays: number | null;
}

interface SessionLogSnapshot {
  logs: string[];
  roots: SessionRootScan[];
}

interface SkillUsageTerms {
  skillId: string;
  skillName: string;
  pathTerms: string[];
}

export interface UsageAnalyzerOptions {
  homeDir: string;
  maxLogBytes?: number;
  maxLogFiles?: number;
}

export class UsageAnalyzer {
  static readonly codexActiveWindowDays = 120;

  readonly homeDir: string;
  readonly maxLogBytes: number;
  readonly maxLogFiles: number;
  private ambiguousMatchesExcluded = 0;
  private timestampFallbackCount = 0;
  private readonly analysisWarningsBySkill = new Map<string, string[]>();

  constructor(options: UsageAnalyzerOptions) {
    this.homeDir = options.homeDir;
    this.maxLogBytes = options.maxLogBytes ?? 512 * 1024;
    this.maxLogFiles = options.maxLogFiles ?? 300;
  }

  async analyzeSkillUsage(skills: SkillRecord[]): Promise<Map<string, UsageHit>> {
    this.ambiguousMatchesExcluded = 0;
    this.timestampFallbackCount = 0;
    this.analysisWarningsBySkill.clear();
    const terms = normalizedPathTerms(skills);
    if (terms.size === 0) return new Map();

    const hits = new Map<string, UsageHit>();
    for (const logPath of await this.sessionLogPaths()) {
      const stat = await fs.promises.stat(logPath).catch(() => null);
      const modifiedAt = stat?.mtime.toISOString() ?? new Date(0).toISOString();
      const matches = await this.matchedSkillEvidence(logPath, terms, modifiedAt);
      for (const [skillId, evidence] of matches) {
        const hit = hits.get(skillId) ?? { count: 0, lastUsedAt: null, evidence: [] };
        hit.count += evidence.length;
        const latestEvidenceAt = evidence
          .map((item) => item.occurredAt)
          .filter((value): value is string => value !== null)
          .sort((a, b) => b.localeCompare(a))[0] ?? null;
        if (latestEvidenceAt && (!hit.lastUsedAt || latestEvidenceAt > hit.lastUsedAt)) hit.lastUsedAt = latestEvidenceAt;
        hit.evidence.push(...evidence);
        hit.evidence.sort((a, b) => (b.occurredAt ?? "").localeCompare(a.occurredAt ?? ""));
        if (hit.evidence.length > 20) hit.evidence = hit.evidence.slice(0, 20);
        hits.set(skillId, hit);
      }
    }
    return hits;
  }

  analysisAudit(): UsageEvidenceAudit {
    const warnings = [...new Set([...this.analysisWarningsBySkill.values()].flat())];
    return {
      ambiguousMatchesExcluded: this.ambiguousMatchesExcluded,
      timestampFallbackCount: this.timestampFallbackCount,
      warnings
    };
  }

  warningsBySkillId(): Map<string, string[]> {
    return new Map(
      [...this.analysisWarningsBySkill].map(([skillId, warnings]) => [skillId, [...warnings]])
    );
  }

  async sessionLogPaths(): Promise<string[]> {
    return (await this.sessionLogSnapshot()).logs;
  }

  async sessionRootAudits(): Promise<UsageSessionRootAudit[]> {
    const snapshot = await this.sessionLogSnapshot();
    return snapshot.roots.map((root) => {
      const logCount = snapshot.logs.filter((logPath) => isPathInside(logPath, root.path)).length;
      return {
        path: root.path,
        agent: root.agent,
        exists: root.exists,
        logCount,
        eligibleLogCount: root.paths.length,
        oversizedLogCount: root.oversizedCount,
        excludedByFileLimitCount: Math.max(0, root.paths.length - logCount),
        timeWindowDays: root.timeWindowDays
      };
    });
  }

  private async matchedSkillEvidence(
    logPath: string,
    terms: Map<string, SkillUsageTerms>,
    modifiedAt: string
  ): Promise<Map<string, UsageEvidence[]>> {
    let raw: string;
    try {
      raw = await fs.promises.readFile(logPath, "utf8");
    } catch {
      return new Map();
    }
    if (!raw.trim()) return new Map();

    const result = new Map<string, UsageEvidence[]>();
    const lines = raw.split(/\r?\n/);
    lines.forEach((line, lineIndex) => {
      if (!isPotentialUsageLine(line)) return;
      let object: unknown;
      try {
        object = JSON.parse(line);
      } catch {
        return;
      }

      const lineMatches = new Map<string, UsageEvidence>();
      const eventOccurredAt = timestampFromEvent(object);
      const occurredAt = eventOccurredAt ?? modifiedAt;
      const timestampSource: UsageEvidence["timestampSource"] = eventOccurredAt ? "event" : "file_mtime";
      const skillToolUse = findSkillToolUse(object);
      if (skillToolUse) {
        const candidates = matchingTermsByName(skillToolUse, terms);
        if (candidates.length === 1) {
          const candidate = candidates[0]!;
          const isClaude = logPath.includes(`${path.sep}.claude${path.sep}projects${path.sep}`);
          const kind: UsageEvidenceKind = isClaude ? "claudeSkillTool" : "codexDirectLoad";
          lineMatches.set(
            candidate.skillId,
            usageEvidence(candidate.skillName, logPath, lineIndex, kind, isClaude ? "claude" : "codex", occurredAt, timestampSource, labelForKind(kind), skillToolUse)
          );
        } else if (candidates.length > 1) {
          this.recordAmbiguousMatch(skillToolUse, candidates);
        }
      }

      for (const event of codexToolSearchTexts(object)) {
        if (!event.text.includes("/") && !event.text.includes("SKILL.md")) {
          const candidates = matchingTermsByName(event.text, terms);
          if (candidates.length === 1) {
            const candidate = candidates[0]!;
            lineMatches.set(
              candidate.skillId,
              usageEvidence(candidate.skillName, logPath, lineIndex, event.kind, event.agent, occurredAt, timestampSource, event.detail, event.text)
            );
          } else if (candidates.length > 1) {
            this.recordAmbiguousMatch(event.text, candidates);
          }
          continue;
        }

        if (!event.text.includes("SKILL.md") && !event.text.includes("/skills/")) continue;
        const candidates = matchingTermsByPath(event.text, terms);
        for (const candidate of candidates) {
          lineMatches.set(
            candidate.skillId,
            usageEvidence(candidate.skillName, logPath, lineIndex, event.kind, event.agent, occurredAt, timestampSource, event.detail, event.text.slice(0, 220))
          );
        }
      }

      if (!eventOccurredAt) this.timestampFallbackCount += lineMatches.size;

      for (const [skillId, evidence] of lineMatches) {
        const current = result.get(skillId) ?? [];
        current.push(evidence);
        result.set(skillId, current);
      }
    });

    return result;
  }

  private recordAmbiguousMatch(observed: string, candidates: SkillUsageTerms[]): void {
    this.ambiguousMatchesExcluded += 1;
    const names = [...new Set(candidates.map((candidate) => candidate.skillName))].join(", ");
    const warning = `Excluded ambiguous usage evidence for "${observed}"; matching installed skills: ${names}. Use path-based evidence to identify a specific copy.`;
    for (const candidate of candidates) {
      const warnings = this.analysisWarningsBySkill.get(candidate.skillId) ?? [];
      if (!warnings.includes(warning)) warnings.push(warning);
      this.analysisWarningsBySkill.set(candidate.skillId, warnings);
    }
  }

  private async sessionLogSnapshot(): Promise<SessionLogSnapshot> {
    const codexActiveRoot = path.join(this.homeDir, ".codex", "sessions");
    const codexArchiveRoot = path.join(this.homeDir, ".codex", "archived_sessions");
    const claudeRoot = path.join(this.homeDir, ".claude", "projects");
    const [codexActive, codexArchive, claude] = await Promise.all([
      this.codexRecentSessionLogs(codexActiveRoot, UsageAnalyzer.codexActiveWindowDays),
      scanShallowLogs(codexArchiveRoot, this.maxLogBytes),
      this.claudeProjectLogs(claudeRoot)
    ]);
    const roots: SessionRootScan[] = [
      {
        ...codexActive,
        path: codexActiveRoot,
        agent: "codex",
        exists: await pathExists(codexActiveRoot),
        timeWindowDays: UsageAnalyzer.codexActiveWindowDays
      },
      {
        ...codexArchive,
        path: codexArchiveRoot,
        agent: "codex",
        exists: await pathExists(codexArchiveRoot),
        timeWindowDays: null
      },
      {
        ...claude,
        path: claudeRoot,
        agent: "claude",
        exists: await pathExists(claudeRoot),
        timeWindowDays: null
      }
    ];

    const unique = [...new Set(roots.flatMap((root) => root.paths))];
    const dated = await Promise.all(
      unique.map(async (logPath) => ({
        logPath,
        mtime: (await fs.promises.stat(logPath).catch(() => null))?.mtimeMs ?? 0
      }))
    );
    const logs = dated
      .sort((a, b) => b.mtime - a.mtime)
      .slice(0, this.maxLogFiles)
      .map((item) => item.logPath);
    return { logs, roots };
  }

  private async codexRecentSessionLogs(root: string, days: number): Promise<ShallowLogScan> {
    const today = new Date();
    const scans: ShallowLogScan[] = [];
    for (let offset = 0; offset < days; offset += 1) {
      const date = new Date(today);
      date.setDate(today.getDate() - offset);
      const directory = path.join(
        root,
        String(date.getFullYear()).padStart(4, "0"),
        String(date.getMonth() + 1).padStart(2, "0"),
        String(date.getDate()).padStart(2, "0")
      );
      scans.push(await scanShallowLogs(directory, this.maxLogBytes));
    }
    return combineScans(scans);
  }

  private async claudeProjectLogs(root: string): Promise<ShallowLogScan> {
    const entries = await fs.promises.readdir(root, { withFileTypes: true }).catch(() => []);
    const projects = entries.filter((entry) => entry.isDirectory());
    const scans = await Promise.all(projects.map((project) => scanShallowLogs(path.join(root, project.name), this.maxLogBytes)));
    return combineScans(scans);
  }
}

async function scanShallowLogs(directory: string, maxLogBytes: number): Promise<ShallowLogScan> {
  const entries = await fs.promises.readdir(directory, { withFileTypes: true }).catch(() => []);
  const paths: string[] = [];
  let oversizedCount = 0;
  for (const entry of entries) {
    if (!entry.isFile()) continue;
    const logPath = path.join(directory, entry.name);
    if (!isSearchableLog(logPath)) continue;
    const stat = await fs.promises.stat(logPath).catch(() => null);
    if (!stat) continue;
    if (stat.size > maxLogBytes) {
      oversizedCount += 1;
      continue;
    }
    paths.push(logPath);
  }
  return { paths, oversizedCount };
}

function combineScans(scans: ShallowLogScan[]): ShallowLogScan {
  return {
    paths: scans.flatMap((scan) => scan.paths),
    oversizedCount: scans.reduce((sum, scan) => sum + scan.oversizedCount, 0)
  };
}

function isPathInside(candidate: string, root: string): boolean {
  const relative = path.relative(root, candidate);
  return relative === "" || (!relative.startsWith("..") && !path.isAbsolute(relative));
}

function normalizedPathTerms(skills: SkillRecord[]): Map<string, SkillUsageTerms> {
  const result = new Map<string, SkillUsageTerms>();
  for (const skill of skills) {
    const skillDirectoryPrefix = `${skill.path}${path.sep}`;
    const variants = new Set([skillDirectoryPrefix, skill.skillFilePath]);
    for (const value of [skillDirectoryPrefix, skill.skillFilePath]) {
      const slashPath = value.replaceAll(path.sep, "/");
      variants.add(slashPath);
      for (const marker of ["/.agents/", "/.codex/", "/.claude/"]) {
        const markerIndex = slashPath.indexOf(marker);
        if (markerIndex >= 0) variants.add(`~${slashPath.slice(markerIndex)}`);
      }
    }
    result.set(skill.id, {
      skillId: skill.id,
      skillName: skill.name,
      pathTerms: [...variants].filter(Boolean).sort((a, b) => b.length - a.length)
    });
  }
  return result;
}

function matchingTermsByName(observedSkillName: string, terms: Map<string, SkillUsageTerms>): SkillUsageTerms[] {
  const normalized = observedSkillName.trim().toLocaleLowerCase();
  if (!normalized) return [];
  return [...terms.values()].filter((term) => term.skillName.toLocaleLowerCase() === normalized);
}

function matchingTermsByPath(observedText: string, terms: Map<string, SkillUsageTerms>): SkillUsageTerms[] {
  const slashText = normalizedEvidencePathText(observedText);
  return [...terms.values()].filter((term) =>
    term.pathTerms.some((candidate) => {
      const slashCandidate = normalizedEvidencePathText(candidate);
      return slashCandidate.length > 0 && slashText.includes(slashCandidate);
    })
  );
}

function normalizedEvidencePathText(value: string): string {
  return value.replaceAll("\\", "/").replace(/\/{2,}/g, "/");
}

function isPotentialUsageLine(line: string): boolean {
  return (
    line.includes('"name":"Skill"') ||
    line.includes('"name": "Skill"') ||
    line.includes('"name":"loadSkill"') ||
    line.includes('"name": "loadSkill"') ||
    line.includes('"name":"load_skill"') ||
    line.includes('"name": "load_skill"') ||
    line.includes("SKILL.md") ||
    line.includes("/skills/")
  );
}

function usageEvidence(
  original: string,
  logPath: string,
  lineIndex: number,
  kind: UsageEvidenceKind,
  agent: SkillAgent,
  occurredAt: string,
  timestampSource: UsageEvidence["timestampSource"],
  detail: string,
  matchedText: string
): UsageEvidence {
  const idSource = `${original}|${logPath}|${lineIndex}|${kind}`;
  return {
    id: stableId(idSource),
    skillName: original,
    agent,
    kind,
    sessionPath: logPath,
    sessionKind: logPath.includes(`${path.sep}.codex${path.sep}archived_sessions`) ? "archived" : "active",
    occurredAt,
    timestampSource,
    detail,
    matchedText,
    confidence: "high"
  };
}

function findSkillToolUse(value: unknown): string | null {
  if (Array.isArray(value)) {
    for (const child of value) {
      const found = findSkillToolUse(child);
      if (found) return found;
    }
    return null;
  }

  if (!value || typeof value !== "object") return null;
  const record = value as Record<string, unknown>;
  if (record.type === "tool_use" && record.name === "Skill" && isRecord(record.input)) {
    const skill = record.input.skill;
    if (typeof skill === "string") return skill;
  }

  for (const child of Object.values(record)) {
    const found = findSkillToolUse(child);
    if (found) return found;
  }
  return null;
}

function codexToolSearchTexts(value: unknown): UsageSearchText[] {
  if (!isRecord(value) || value.type !== "response_item" || !isRecord(value.payload)) return [];
  const payload = value.payload;
  if (payload.type !== "function_call" && payload.type !== "custom_tool_call") return [];
  const name = typeof payload.name === "string" ? payload.name : "";

  if (isDirectSkillToolName(name)) {
    const skillName = skillNameFromToolPayload(payload);
    return skillName ? [{ text: skillName, kind: "codexDirectLoad", agent: "codex", detail: `Codex ${name}` }] : [];
  }

  if (!["exec_command", "read_mcp_resource", "open"].includes(name)) return [];
  return stringValues(payload, ["arguments", "input"]).map((text) => ({
    text,
    kind: "codexSkillRead",
    agent: "codex",
    detail: `Codex ${name} read`
  }));
}

function isDirectSkillToolName(name: string): boolean {
  return ["Skill", "loadSkill", "load_skill"].includes(name);
}

function skillNameFromToolPayload(payload: Record<string, unknown>): string | null {
  for (const value of stringValues(payload, ["arguments", "input"])) {
    try {
      const object = JSON.parse(value) as unknown;
      const skill = skillNameFromJsonObject(object);
      if (skill) return skill;
    } catch {
      const trimmed = value.trim();
      if (trimmed && !trimmed.includes("/")) return trimmed;
    }
  }
  return null;
}

function skillNameFromJsonObject(value: unknown): string | null {
  if (!isRecord(value)) return null;
  for (const key of ["skill", "skillName", "name"]) {
    const item = value[key];
    if (typeof item === "string") return item;
  }
  return null;
}

function stringValues(record: Record<string, unknown>, keys: string[]): string[] {
  return keys.flatMap((key) => {
    const value = record[key];
    if (typeof value === "string") return [value];
    if (value === undefined) return [];
    try {
      return [JSON.stringify(value)];
    } catch {
      return [];
    }
  });
}

function timestampFromEvent(value: unknown): string | null {
  if (!isRecord(value)) return null;
  for (const key of ["timestamp", "time", "createdAt", "created_at"]) {
    const timestamp = normalizedTimestamp(value[key]);
    if (timestamp) return timestamp;
  }
  for (const key of ["payload", "message", "event"]) {
    if (isRecord(value[key])) {
      const timestamp = timestampFromEvent(value[key]);
      if (timestamp) return timestamp;
    }
  }
  return null;
}

function normalizedTimestamp(value: unknown): string | null {
  if (typeof value === "number") {
    if (!Number.isFinite(value) || value < 1_000_000_000) return null;
    return isoFromDate(value < 1_000_000_000_000 ? value * 1000 : value);
  }
  if (typeof value !== "string" || !value.trim()) return null;
  const numeric = Number(value);
  if (Number.isFinite(numeric) && value.trim().match(/^\d+(?:\.\d+)?$/)) {
    return normalizedTimestamp(numeric);
  }
  return isoFromDate(value);
}

function labelForKind(kind: UsageEvidenceKind): string {
  switch (kind) {
    case "codexSkillRead":
      return "Codex read SKILL.md";
    case "codexDirectLoad":
      return "Codex loadSkill";
    case "claudeSkillTool":
      return "Claude Skill tool";
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value && typeof value === "object" && !Array.isArray(value));
}
